src/config.rs

```diff
@@ -110,6 +116,112 @@ configuration_option_enum! { TypeDensity:
     Wide,
 }
 
+// Newtype needed to have impls for `create_config` macro.
+#[derive(Clone, Debug, Default, RustcDecodable)]
+pub struct FileLinesMap(pub HashMap<String, LineSet>);
+
+impl ConfigType for Option<FileLinesMap> {
+    type ParseErr = String;
+
+    fn doc_hint() -> String {
+        unimplemented!();
+    }
+
+    fn parse(s: &str) -> Result<Self, String> {
+        let mut map = HashMap::new();
+        let (file, lines) = try!(parse_file_lines_spec(s));
+        map.insert(file, lines);
+
+        Ok(Some(FileLinesMap(map)))
+    }
+}
+
+pub fn parse_file_lines_spec(s: &str) -> Result<(String, LineSet), String> {
+    let err = || Err(format!("invalid experimental-file-lines argument: {}", s));
+
+    match file_lines_spec_parser::file_lines_spec(&s) {
+        IResult::Error(_) |
+        IResult::Incomplete(_) => return err(),
+        IResult::Done(remaining, _) if !remaining.is_empty() => return err(),
+        IResult::Done(_, (file, line_ranges)) => Ok((file, line_ranges)),
+    }
+}
+
+impl iter::FromIterator<(String, LineSet)> for FileLinesMap {
+    fn from_iter<I: IntoIterator<Item = (String, LineSet)>>(iter: I) -> FileLinesMap {
+        let mut ret = FileLinesMap(HashMap::new());
+        ret.extend(iter);
+        ret
+    }
+}
+
+impl Extend<(String, LineSet)> for FileLinesMap {
+    fn extend<I>(&mut self, iter: I)
+        where I: IntoIterator<Item = (String, LineSet)>
+    {
+        let map = &mut self.0;
+
+        for (file, line_set) in iter {
+            map.entry(file).or_insert(LineSet::new()).extend(line_set);
+        }
+    }
+}
+
+mod file_lines_spec_parser {
```

Comment:
> I don't really see why this is a map rather than a list (internally it might need to be a map, but in the API it's just a list of spans, no?)

That's a interesting thought. It *is* a list of spans, but provided before we have parsed anything. The way control passes at the moment, we can't resolve the spans in the codemap early enough, since the `rustfmt::run` handles the parsing and accepts a resolved config. I'll take a look at flow around there to see if we can resolve everything to spans, which would potentially let me get rid of the whole `LineSet` thing.